kube-proxy: Startup and updates logging

thockin opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

Improve logging at startup and in steady-state kube-proxy.  A normal logstream now looks like:

```
I0409 20:52:01.661124 1755549 endpoints.go:181] Service port default/hostnames-clip:p80 updated: 1 endpoints
I0409 20:52:01.661204 1755549 endpoints.go:181] Service port default/hostnames-clip:p81 updated: 1 endpoints
I0409 20:52:01.661946 1755549 proxier.go:840] Syncing iptables rules
I0409 20:52:01.756777 1755549 proxier.go:809] syncProxyRules took 95.441782ms
I0409 20:52:37.830382 1755549 endpoints.go:181] Service port default/hostnames-clip:p80 updated: 2 endpoints
I0409 20:52:37.830489 1755549 endpoints.go:181] Service port default/hostnames-clip:p81 updated: 2 endpoints
I0409 20:52:37.831309 1755549 proxier.go:840] Syncing iptables rules
I0409 20:52:37.975715 1755549 proxier.go:809] syncProxyRules took 145.06061ms
I0409 20:52:53.707041 1755549 endpoints.go:181] Service port default/hostnames-clip:p80 updated: 2 endpoints
I0409 20:52:53.707084 1755549 endpoints.go:181] Service port default/hostnames-clip:p81 updated: 2 endpoints
I0409 20:52:53.707647 1755549 proxier.go:840] Syncing iptables rules
I0409 20:52:53.847483 1755549 proxier.go:809] syncProxyRules took 140.357085ms
I0409 20:53:36.058767 1755549 endpoints.go:181] Service port default/hostnames-clip:p80 updated: 2 endpoints
I0409 20:53:36.058817 1755549 endpoints.go:181] Service port default/hostnames-clip:p81 updated: 2 endpoints
I0409 20:53:36.059386 1755549 proxier.go:840] Syncing iptables rules
I0409 20:53:36.102663 1755549 proxier.go:809] syncProxyRules took 43.807718ms
I0409 20:53:58.986360 1755549 service.go:267] Service default/hostnames-clip updated: 2 ports
I0409 20:53:58.986985 1755549 service.go:383] Updating existing service port "default/hostnames-clip:p80" at 10.1.0.209:80/TCP
I0409 20:53:58.987090 1755549 service.go:381] Adding new service port "default/hostnames-clip:p82" at 10.1.0.209:82/TCP
I0409 20:53:58.987113 1755549 service.go:406] Removing service port "default/hostnames-clip:p81"
I0409 20:53:58.987181 1755549 proxier.go:840] Syncing iptables rules
I0409 20:53:58.993265 1755549 endpoints.go:181] Service port default/hostnames-clip:p82 updated: 2 endpoints
I0409 20:53:58.993296 1755549 endpoints.go:181] Service port default/hostnames-clip:p80 updated: 2 endpoints
I0409 20:53:59.043839 1755549 proxier.go:809] syncProxyRules took 57.421966ms
I0409 20:53:59.044477 1755549 proxier.go:840] Syncing iptables rules
I0409 20:53:59.080437 1755549 proxier.go:809] syncProxyRules took 36.551818ms
I0409 20:54:38.021088 1755549 service.go:267] Service default/hostnames-clip updated: 1 ports
I0409 20:54:38.021721 1755549 service.go:381] Adding new service port "default/hostnames-clip" at 10.1.0.209:80/TCP
I0409 20:54:38.021828 1755549 service.go:406] Removing service port "default/hostnames-clip:p80"
I0409 20:54:38.021891 1755549 service.go:406] Removing service port "default/hostnames-clip:p82"
I0409 20:54:38.022009 1755549 proxier.go:840] Syncing iptables rules
I0409 20:54:38.030413 1755549 endpoints.go:181] Service port default/hostnames-clip updated: 2 endpoints
I0409 20:54:38.269695 1755549 proxier.go:809] syncProxyRules took 248.430424ms
I0409 20:54:38.270226 1755549 proxier.go:840] Syncing iptables rules
I0409 20:54:38.311005 1755549 proxier.go:809] syncProxyRules took 41.26259ms
```


```release-note
NONE
```

thockin committed:
kube-proxy: more logging at startup

thockin committed:
kube-proxy: log service and endpoint updates

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -273,6 +273,7 @@ func NewProxier(ipt utiliptables.Interface,
 		if err := sysctl.SetSysctl(sysctlRouteLocalnet, 1); err != nil {
 			return nil, fmt.Errorf("can't set sysctl %s: %v", sysctlRouteLocalnet, err)
 		}
+		klog.V(1).Info("Set sysctl %q to 1", sysctlRouteLocalnet)
```
maybe just make `sysctl.SetSysctl` log this itself?

thockin replied:
I don't like logging from deep within libraries like that, but I can extract to proxy/util

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -319,18 +321,35 @@ func NewProxier(ipt utiliptables.Interface,
 	burstSyncs := 2
-	klog.V(3).Infof("minSyncPeriod: %v, syncPeriod: %v, burstSyncs: %d", minSyncPeriod, syncPeriod, burstSyncs)
+	klog.V(2).Infof("iptables(%s) sync params: minSyncPeriod=%v, syncPeriod=%v, burstSyncs=%d",
+		ipVersion(ipt.IsIpv6()), minSyncPeriod, syncPeriod, burstSyncs)
```
it might be nice if all of the startup configuration logging was all together?

thockin replied:
I was trying to log close to where the things were relevant - these params most influence the BoundedFrequency logic..

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -319,18 +321,35 @@ func NewProxier(ipt utiliptables.Interface,
 		proxier.syncProxyRules, syncPeriod, wait.NeverStop)
+
+	if ipt.HasRandomFully() {
+		klog.V(2).Infof("iptables(%s) supports --random-fully", ipVersion(ipt.IsIpv6()))
```
Just when I thought I was out (#89296), they pull me back in!

thockin replied:
Had a customer today who wanted to know from logs.  Logging once as opposed to every sync, seems like a good tradeoff?

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -319,18 +321,35 @@ func NewProxier(ipt utiliptables.Interface,
 	return proxier, nil
 }
 
+func ipVersion(isIPv6 bool) string {
```
`pkg/util/iptables/iptables_test.go` has a `protocolStr` function that does basically this, and maybe you should just make it an `iptables.Interface` function and use it from here and there?

thockin replied:
Great idea.  Next push does this and some extra cleanup (review commit-by-commit)

danwinship reviewed (approved):
I've just got nitpicks. It could be lgtm.

thockin committed:
Add Protocol() method to iptables

Enables simpler printing of which IP family the iptables interface is
managing.

thockin committed:
Rename iptables IsIpv6 to IsIPv6

thockin commented:
/retest

thockin committed:
Encapsulate sysctl test and log

thockin committed:
Cleanup: Change "Ip" to "IP" in func and var names

thockin commented:
Oh, I have to fix up "golden" state because of the `<unnamed>` change - any thoughts on whether this is better than an empty string port name?

danwinship commented on pkg/kubelet/dockershim/network/hostport/hostport_manager_test.go:
```diff
@@ -387,7 +387,7 @@ func TestGetHostportChain(t *testing.T) {
 func TestHostportManagerIPv6(t *testing.T) {
 	iptables := NewFakeIPTables()
-	iptables.ipv6 = true
+	iptables.protocol = utiliptables.ProtocolIPv4
```
`ProtocolIPv6`
(that's the other breakage in pull-kubernetes-bazel-test)

thockin replied:
ACk

danwinship reviewed (commented):
`<unnamed>` doesn't really seem like it clarifies anything. Maybe just drop the `:` if there's no name?

everything else looks good

thockin commented:
I was really torn on <unnamed> or just dropping the colon.  Since you landed on that side, I'll do that. :)

thockin commented:
Re-pushed with changes suggested and tests fixed

danwinship commented:
/lgtm
